connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala

```diff
@@ -0,0 +1,131 @@
+    KVUtils.viewToSeq(store.view(classOf[SessionInfo]))
+  }
+
+  def getExecutionList: Seq[ExecutionInfo] = {
+    KVUtils.viewToSeq(store.view(classOf[ExecutionInfo]))
+  }
+
+  def getOnlineSessionNum: Int = {
+    KVUtils.count(store.view(classOf[SessionInfo]))(_.finishTimestamp == 0)
+  }
+
+  def getSession(sessionId: String): Option[SessionInfo] = {
+    try {
+      Some(store.read(classOf[SessionInfo], sessionId))
+    } catch {
+      case _: NoSuchElementException => None
+    }
+  }
+
+  def getExecution(executionId: String): Option[ExecutionInfo] = {
+    try {
+      Some(store.read(classOf[ExecutionInfo], executionId))
+    } catch {
+      case _: NoSuchElementException => None
+    }
+  }
+
+  /**
+   * When an error or a cancellation occurs, we set the finishTimestamp of the statement.
+   * Therefore, when we count the number of running statements, we need to exclude errors and
+   * cancellations and count all statements that have not been closed so far.
+   */
+  def getTotalRunning: Int = {
+    KVUtils.count(store.view(classOf[ExecutionInfo]))(_.isExecutionActive)
+  }
+
+  def getSessionCount: Long = {
+    store.count(classOf[SessionInfo])
+  }
+
+  def getExecutionCount: Long = {
+    store.count(classOf[ExecutionInfo])
+  }
+}
+
+private[connect] class SessionInfo(
+    @KVIndexParam val sessionId: String,
+    val startTimestamp: Long,
+    val userId: String,
+    val finishTimestamp: Long,
+    val totalExecution: Long) {
+  @JsonIgnore @KVIndex("finishTime")
+  private def finishTimeIndex: Long = if (finishTimestamp > 0L) finishTimestamp else -1L
+  def totalTime: Long = {
+    if (finishTimestamp == 0L) {
+      System.currentTimeMillis - startTimestamp
+    } else {
+      finishTimestamp - startTimestamp
+    }
+  }
+}
+
+private[connect] class ExecutionInfo(
+    @KVIndexParam val jobTag: String,
+    val statement: String,
+    val sessionId: String,
+    val startTimestamp: Long,
+    val userId: String,
+    val operationId: String,
+    val sparkSessionTags: Set[String],
+    val finishTimestamp: Long,
+    val closeTimestamp: Long,
+    val detail: String,
+    val state: ExecutionState.Value,
+    val jobId: ArrayBuffer[String],
+    val sqlExecId: mutable.Set[String]) {
+  @JsonIgnore @KVIndex("finishTime")
+  private def finishTimeIndex: Long = if (finishTimestamp > 0L && !isExecutionActive) {
+    finishTimestamp
+  } else -1L
+
+  @JsonIgnore @KVIndex("isExecutionActive")
+  def isExecutionActive: Boolean = {
+    state == ExecutionState.STARTED ||
+    state == ExecutionState.COMPILED ||
+    state == ExecutionState.READY
+  }
+
+  def totalTime(endTime: Long): Long = {
+    if (endTime == 0L) {
+      System.currentTimeMillis - startTimestamp
+    } else {
+      endTime - startTimestamp
+    }
+  }
+}
+
+private[connect] object ExecutionState extends Enumeration {
+  val STARTED, COMPILED, READY, CANCELED, FAILED, FINISHED, CLOSED = Value
+  type ExecutionState = Value
+}
```

Comment:
For potential followup:
I find it a bit weird, that it doesn't have a RUNNING execution state, but it's actually the same as in Thriftserver:
https://github.com/apache/spark/blob/master/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala#L119C58-L119C58

Spark Connect records that it's "ready for execution" at the end of planner (or before eager commands), and we record "finished" when the result stage completes... Should we have another event "RUNNING" that e.g. for SparkConnectPlanExecution would be right before `spark.sparkContext.submitJob`, and for Commands would need to be determined on a per command basis. WDYT @jdesjean ; that would be different from Thriftserver, but it seems to me like that "RUNNING" state is missing from Thriftserver... cc @wangyum @yaooqinn  @bogdanghit WDYT from the Thriftserver perspective?

Reply:
I think a `RUNNING` state should cover the period between it enters `STARTED` state until it enters one of the terminal states. Using this definition, in ThriftServer we can infer the execution is in running / active state if it's not `FAILED, CANCELED, TIMEDOUT or CLOSED.`

Nit: It would be good to keep these states consistent. The SparkConnect execution state has a `READY` state which is missing in ThriftServer.

Reply:
@bogdanghit see https://github.com/apache/spark/pull/41748 from @jdesjean . Jean did some work to be able to figure out the "READY" state to come before the start of eager commands. In Thriftserver there's a bit of ugly issue that eager commands would execute even before the COMPILED state, because they are eagerly executed straight in analysis.
Thriftserver could use https://github.com/apache/spark/pull/41748 to also display information about eager execution better.

Reply:
I have the UI display the state as RUNNING is isActiveExecution is true, which covers STARTED, COMPILED, READY: https://github.com/apache/spark/pull/41964/commits/8abdecdc7cedb6ff5a13090dd4bcae2ecf50d0d0#diff-ef2b5b937c2c5e83d6ca9804c9d402d263bf802550cc8daeead089498c123aa0L322.